### ISSUE TITLE
[Sept Sprint] New exercise teaching pointers

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/go#1643]
This is part of our September Spring to get the Go Track in shape. Please read the [Overview issue](https://github.com/exercism/go/issues/1610) for more details. **This exercise is unclaimed. To claim it, please comment below.**

---

**Before building this exercise (or exercises), please read up on [the documentation](https://exercism.org/docs/building/tracks/concept-exercises).**

We need an exercise for the `pointers` and `nil` concepts. These concepts go well together so it makes sense to teach them in the same exercise.

The exercise should teach:

- What pointers are
- Pointer types for primitives
- Which types are already pointers
- Explain that go has no pointer arithmetic
- Referencing and dereferencing syntax
- `nil` as zero value for the pointer
- pitfalls when working with pointers (always check whether they are nil before dereferencing)

Keep in mind that the concept of pointers might be hard for people coming from languages that don't have explicit pointer types.

**Comment by junedev** [comment]
We have noticed in the exercise teaching `zero-values` that it might not make sense to create a separate concept for `nil`. You can also just include the info in the `pointers` concept instead and have this exercise just teach `pointers`.

**Comment by ErikSchierboom** [comment]
> have this exercise just teach `pointers`.

Sounds like that might be better actually, seeing how people often struggle with pointers.

**Comment by andrerfcsantos** [comment]
Just to be clear, this actually requires the implementation of a concept + exercise, right?

If so, what would be the requirements for the concept?

**Comment by junedev** [comment]
Yes, exactly. The introduction file of the exercise should cover the list that you find in the issue description. Then you can use that same content as introduction file of the concept. You can also copy the content into about.md of the concept and then think about whether you would like to add more detailed content/examples there etc. Also if some item on the issue description list is not needed in the exercise and not needed to understand the concept in general, then it should be moved to that about.md file of the concept.

Besides introduction and about there is not much needed for a concept, you can find the general documentation on Concepts here https://exercism.org/docs/building/tracks/concepts. 

TLDR If you focus on the exercise and its introduction, you have all the minimal things you need for the concept as well.

**Comment by ErikSchierboom** [comment]
@andrerfcsantos I'm in the process of adding a new exercise, so you can check my PR to give you an idea of what it looks like: https://github.com/exercism/go/pull/1680 Note that it's perfectly fine to create the exercise in a series of smaller, incremental PRs if you'd like that.

**Comment by andrerfcsantos** [comment]
@ErikSchierboom @junedev Thanks for the pointers. But I think I was not clear enough in my previous message, sorry for that. What I meant by "requirements"  is what concepts should come before this one, i.e what should we assume the student knows before getting to this concept. This would be relevant to know which examples of pointers we can include.

**Comment by junedev** [comment]
@andrerfcsantos 🙈 Sorry, totally misunderstood the question.

You can and should put "structs", "maps" and "slices" as prerequisites for this exercise so that you can talk about things like a slice already being a pointer etc. Also the student already knows about the primitive types of course. The student will learn about methods and pointer receivers AFTER this exercise so you should not go into that. If there are other concepts you feel you need as prerequisite and are unsure about, just leave me another comment.

**Comment by andrerfcsantos** [comment]
Since no one seems to be claiming this one and we are near Oct 1, I'll work on this one.

**Comment by ErikSchierboom** [comment]
@andrerfcsantos Let me know if there is anything I can do to help.

**Comment by andrerfcsantos** [comment]
While making this exercise, I noticed something that we must also add to the exercise, and that is to talk about pointers to structs. When we have a pointer to a struct in Go, we don't need to dereference it to access its fields - Go does that automatically. I think it'd be important to show this.

**Comment by ErikSchierboom** [comment]
That sounds like a good idea. There exercise with the `structs` concept  has only two prerequisites: `conditionals-if` and `numbers`, so I think you can safely add `structs` as a prerequisite. @junedev?

**Comment by junedev** [comment]
@ErikSchierboom Do you mean adding `pointers` as prerequisite?
I don't think that is necessary, it makes sense that structs are available rather early in the track. What @andrerfcsantos mentioned should be part of this exercise/concept (and this is also what he suggested I think). You don't need that information before you know how to create a pointer to a struct which is covered here.

**Comment by ErikSchierboom** [comment]
@junedev I meant the opposite. `structs` as a prerequisite of `pointers`

**Comment by junedev** [comment]
Yes, thats what I wrote above.
> You can and should put "structs", "maps" and "slices" as prerequisites

**Comment by ErikSchierboom** [comment]
Ehm, you wrote:

> @ErikSchierboom Do you mean adding `pointers` as prerequisite for the `structs` exercise?

Or am I misreading? :D

Anyways, we both want to have `structs` to be a prerequisite of `pointers`.

**Comment by junedev** [comment]
I meant further above https://github.com/exercism/go/issues/1643#issuecomment-925771557

**Comment by ErikSchierboom** [comment]
Ah right

**Comment by andrerfcsantos** [comment]
When writing the `design.md` for the exercise I noticed we haven't discussed what concepts should this unlock.

When looking at the current tree I don't immediate candidates. Maybe this will unlock `functions` when that concept is available? Should this not have any unlocks for now?

**Comment by junedev** [comment]
@andrerfcsantos The "this exercise unlocks ..." part in the design file should list the concepts that are included with the exercise, so `pointers` in your case. You don't have to think about what comes after your exercise in the concept tree.

**Comment by andrerfcsantos** [comment]
@junedev Ah, totally misunderstood it. Thanks for clarifying.

**Comment by junedev** [comment]
Yes, I meant maps and slices. You might be right this deserves another exercise but it might take quite some time until someone gets around to creating one. So I would recommend at least adding a short bit on this here for now. It can be moved/removed later.

**Comment by andrerfcsantos** [comment]
Ready to review #1724

I might to take another look at it tomorrow, but getting some feedback would be nice. This is my first exercise and I'm sure there are things to improve :)
